canal/sync.go

```diff
@@ -65,6 +65,10 @@ func (c *Canal) runSyncBinlog() error {
 		// TODO: If we meet any DDL query, we must save too.
 		switch e := ev.Event.(type) {
 		case *replication.RotateEvent:
+			// if event pos equal to local pos.maybe be fake rotate event,do nothing.
+			if string(e.NextLogName) == pos.Name && uint32(e.Position) == pos.Pos {
```

Comment:
in MySQL, a fake rotate event is an event with _timestamp_ equals to **0** (see <https://github.com/mysql/mysql-server/blob/8cc757da3d87bf4a1f07dcfb2d3c96fed3806870/sql/rpl_binlog_sender.cc#L899>), and also with _log pos_ equals to **0** in the implementation (see <https://github.com/mysql/mysql-server/blob/8cc757da3d87bf4a1f07dcfb2d3c96fed3806870/sql/rpl_binlog_sender.cc#L906>).

I don't know whether some corner cases exist if checking by `string(e.NextLogName) == pos.Name && uint32(e.Position) == pos.Pos`.

Reply:
Friendly ping @lintanghui

Reply:
> in MySQL, a fake rotate event is an event with _timestamp_ equals to **0** (see https://github.com/mysql/mysql-server/blob/8cc757da3d87bf4a1f07dcfb2d3c96fed3806870/sql/rpl_binlog_sender.cc#L899), and also with _log pos_ equals to **0** in the implementation (see https://github.com/mysql/mysql-server/blob/8cc757da3d87bf4a1f07dcfb2d3c96fed3806870/sql/rpl_binlog_sender.cc#L906).
> 
> I don't know whether some corner cases exist if checking by `string(e.NextLogName) == pos.Name && uint32(e.Position) == pos.Pos`.

yes, timestamp 0 and pos 0 is set in event header, but in event body, pos and file name is also be set,   https://github.com/mysql/mysql-server/blob/8cc757da3d87bf4a1f07dcfb2d3c96fed3806870/sql/rpl_binlog_sender.cc#L909  . in sync.go we just get event body and compare it.  we can also do such below
`ev.Header.Pos==0 {continue }`

Reply:
@lintanghui I know, in this special case `string(e.NextLogName) == pos.Name && uint32(e.Position) == pos.Pos` is right, but I don't know whether some **corner cases** exist.

BTW, will you update it to `ev.Header.Pos==0 {continue }`?

Reply:
fine, it seem better to compare header pos.